### ISSUE TITLE
Inadmissible bootstrap estimates when `se = "boot"`

### ISSUE THREAD

**Issue by sfcheung** [yrosseel/lavaan#235]
It seems that `lavaan()` does not check whether the estimates from a bootstrap sample is admissible when `se = "boot"`:

```r
library(lavaan)
#> This is lavaan 0.6-11
#> lavaan is FREE software! Please report any bugs.

check_post_check <- function(x, fit) {
    p_table <- lavaan::parameterTable(fit)
    p_table[p_table$free > 0, "est"] <- x[p_table$free]
    tmp <- lavaan::lavaan(p_table, do.fit = FALSE, check.start = TRUE)
    lavaan::lavInspect(tmp, "post.check")
  }

HS.model <- ' visual  =~ x1 + x2 + x3
              textual =~ x4 + x5 + x6
              speed   =~ x7 + x8 + x9 '

set.seed(58742)
fit_cfa <- cfa(HS.model,
           data = HolzingerSwineford1939,
           se = "boot", bootstrap = 2000)

boot_est <- lavInspect(fit_cfa, "boot")
out <- sapply(split(boot_est, row(boot_est)),
              check_post_check, fit = fit_cfa)
#> Warning in lav_object_post_check(object): lavaan WARNING: some estimated ov
#> variances are negative

#> Warning in lav_object_post_check(object): lavaan WARNING: some estimated ov
#> variances are negative

#> Warning in lav_object_post_check(object): lavaan WARNING: some estimated ov
#> variances are negative

#> Warning in lav_object_post_check(object): lavaan WARNING: some estimated ov
#> variances are negative

#> Warning in lav_object_post_check(object): lavaan WARNING: some estimated ov
#> variances are negative

#> Warning in lav_object_post_check(object): lavaan WARNING: some estimated ov
#> variances are negative

#> Warning in lav_object_post_check(object): lavaan WARNING: some estimated ov
#> variances are negative

#> Warning in lav_object_post_check(object): lavaan WARNING: some estimated ov
#> variances are negative

#> Warning in lav_object_post_check(object): lavaan WARNING: some estimated ov
#> variances are negative

#> Warning in lav_object_post_check(object): lavaan WARNING: some estimated ov
#> variances are negative

#> Warning in lav_object_post_check(object): lavaan WARNING: some estimated ov
#> variances are negative

#> Warning in lav_object_post_check(object): lavaan WARNING: some estimated ov
#> variances are negative

#> Warning in lav_object_post_check(object): lavaan WARNING: some estimated ov
#> variances are negative

#> Warning in lav_object_post_check(object): lavaan WARNING: some estimated ov
#> variances are negative
all(out)
#> [1] FALSE
table(out)
#> out
#> FALSE  TRUE 
#>    14  1986
```

The call of `cfa()` with `se = "boot"` does not return any error message nor warning. However, when I check whether the bootstrap estimates are admissible, some of them are not (14 in the example above).

I verified my `check_post_check()` function using `bootstrapLavaan()`:

```r
fit_cfa_noboot <- update(fit_cfa, se = "none")
set.seed(58742)
boot_est2 <- bootstrapLavaan(fit_cfa_noboot, R = 2000)
out2 <- sapply(split(boot_est2, row(boot_est2)),
              check_post_check, fit = fit_cfa)
#> Warning in lav_object_post_check(object): lavaan WARNING: some estimated ov
#> variances are negative

#> Warning in lav_object_post_check(object): lavaan WARNING: some estimated ov
#> variances are negative

#> Warning in lav_object_post_check(object): lavaan WARNING: some estimated ov
#> variances are negative

#> Warning in lav_object_post_check(object): lavaan WARNING: some estimated ov
#> variances are negative

#> Warning in lav_object_post_check(object): lavaan WARNING: some estimated ov
#> variances are negative

#> Warning in lav_object_post_check(object): lavaan WARNING: some estimated ov
#> variances are negative

#> Warning in lav_object_post_check(object): lavaan WARNING: some estimated ov
#> variances are negative

#> Warning in lav_object_post_check(object): lavaan WARNING: some estimated ov
#> variances are negative

#> Warning in lav_object_post_check(object): lavaan WARNING: some estimated ov
#> variances are negative

#> Warning in lav_object_post_check(object): lavaan WARNING: some estimated ov
#> variances are negative

#> Warning in lav_object_post_check(object): lavaan WARNING: some estimated ov
#> variances are negative

#> Warning in lav_object_post_check(object): lavaan WARNING: some estimated ov
#> variances are negative

#> Warning in lav_object_post_check(object): lavaan WARNING: some estimated ov
#> variances are negative

#> Warning in lav_object_post_check(object): lavaan WARNING: some estimated ov
#> variances are negative
all(out2)
#> [1] FALSE
table(out2)
#> out2
#> FALSE  TRUE 
#>    14  1986
```

The bootstrap estimates are virtually identical (they started with the same random seed), and `bootstrapLavaan()` issued the warning messages when doing bootstrapping.

Although the proportion of inadmissible solution may be small, it still sounds strange to include them when forming the bootstrap confidence interval. Things can become very weird in some situations. I was actually led to examine this issue because I noticed something weird in a data set when I did bootstrapping in lavaan.

May I suggest checking the solution from each bootstrap sample and reject the sample if the solution is not admissible? I drafted the following changes and verified that the 14 samples above will be excluded (i.e., the final number of bootstrap samples is 1986, not the requested 2000):

https://github.com/sfcheung/lavaan/commit/cc819082f41182fb81d257a8ffea06d71cf77ff0

Basically, I just added a test similar to that for nonconvergence, and use the internal function used to "post.check" in `lavInsepct()`.

**Comment by sfcheung** [comment]
I believe I made a mistake. Not that inadmissible solution should be rejected, but that the issue is not that simple. 

First, what to do with inadmissible solution in a bootstrap sample depends on what the sampling distribution of the estimates is.

If, when we fit the model to any possible sample from the same population, the estimates will always be accepted even if at least one of the estimates is inadmissible, assuming there is no problem in minimization, then, to estimate the sampling distribution of the estimates found this way, inadmissible solutions should be retained.

However, this is not what we do in practice. Inadmissible solution is rarely accepted and presented (though mechanically discarding this solution may not be a good practice: see Kolenikov & Bollen, 2012). If this is how we actually find the estimates, with inadmissible solution discarded and hence the sampling distribution does not include inadmissible solutions, then the bootstrapping process should simulate this and reject inadmissible solutions.

Second, a solution may be inadmissible for many different reasons, and there may be solutions other than discarding or retaining it. When we fit a model to one sample, if the solution is inadmissible, we can do what Kolenikov and Bollen (2012) proposed. We can also try things like changing the starting values, and see if we can come up with an admissible solution. The minimization may achieve convergence but there is another solution with a smaller value for the discrepancy function. However, in bootstrapping, it is not easy to do this automatically, even though the number of bootstrap samples with inadmissible solution is small.

Third, a solution may not be just inadmissible but also "unusual". Actually, I explored this issue because I noticed something's odd in the example I presented above. This is the plot of bootstrap estimates of the variances in that example:

```r
# Run the code in the first post of this issue first
library(psych)
pairs.panels(boot_est[, c("visual~~visual", "textual~~textual", "speed~~speed")])
```

![image](https://user-images.githubusercontent.com/7019263/164128746-ae7b8b25-0792-4b0c-88b1-dceca6450e7d.png)


In a few of the bootstrap samples, the estimates of some parameters can be very extreme, and the solution is inadmissible partly because of these "admissible" estimates. For example, these are the estimates of the bootstrap sample with the largest estimate of the variance of `visual`:

```r
# Run the code in the first post of this issue first
boot_est[!out, ][10, ]
#>> boot_est[!out, ][10, ]
#>     visual=~x2       visual=~x3      textual=~x5      textual=~x6
#>      0.1642398        0.2950206        1.0345905        0.8557674
#>       speed=~x8        speed=~x9           x1~~x1           x2~~x2
#>      1.0950900        0.7584166       -0.7347723        1.4466962
#>         x3~~x3           x4~~x4           x5~~x5           x6~~x6
#>       1.0851230        0.3415560        0.4851786        0.3059965
#>         x7~~x7           x8~~x8           x9~~x9   visual~~visual
#>       0.6670267        0.4687220        0.7002756        2.0680230
#>textual~~textual     speed~~speed  visual~~textual    visual~~speed
#>       1.0399275        0.4604670        0.5221166        0.2127842
#>  textual~~speed
#>       0.1217868
```

Therefore, the problem with an inadmissible solution in a bootstrap sample may not be just about negative variances, but also unusual but "admissible" estimates in some other parameters.

I don't have a quick solution. Maybe, when `se = "boot"`, `lavaan()` and/or `lav_bootstrap_internal()` checks the number of inadmissible solutions and report them, and also allows users to decide whether inadmissible solutions should be rejected when the bootstrap estimates are used to compute the standard errors and form the confidence intervals?

Reference:
Kolenikov, S., & Bollen, K. (2012). Testing negative error variances: Is a Heywood case a symptom of misspecification? *Sociological Methods & Research, 41*(1), 124–167.

Note:  Discarding inadmissible solutions, like those with estimates of variances that are negative, can also be problematic for confidence intervals, because the lower bound of the percentile confidence interval of a variance will nearly always be higher then zero, resulting in suboptimal coverage probability when the population value of this variance is small relative to its standard error.

**Comment by sfcheung** [comment]
Setting `bounds` to one of the options (other than `"none"`, the default) may solve the problem:

```r
library(lavaan)
#> This is lavaan 0.6-10
#> lavaan is FREE software! Please report any bugs.

check_post_check <- function(x, fit) {
    p_table <- lavaan::parameterTable(fit)
    p_table[p_table$free > 0, "est"] <- x[p_table$free]
    tmp <- lavaan::lavaan(p_table, do.fit = FALSE, check.start = TRUE)
    lavaan::lavInspect(tmp, "post.check")
  }

HS.model <- ' visual  =~ x1 + x2 + x3
              textual =~ x4 + x5 + x6
              speed   =~ x7 + x8 + x9 '

set.seed(58742)
fit_cfa <- cfa(HS.model,
           data = HolzingerSwineford1939,
           se = "boot", bootstrap = 2000,
           bounds = "standard")

boot_est <- lavInspect(fit_cfa, "boot")
out <- sapply(split(boot_est, row(boot_est)),
              check_post_check, fit = fit_cfa)
all(out)
#> [1] TRUE
table(out)
#> out
#> TRUE 
#> 2000
```

However, users may not be aware of inadmissible estimates in some bootstrap estimates if they just set `se = "boot"`.

Therefore, I think `lav_bootstrap_internal()` should still do something to alert the users of inadmissible solutions in some bootstrap estimates. For example, instead of rejecting them (as proposed in cc819082f41182fb81d257a8ffea06d71cf77ff0), it can set a flag when one or more solutions are inadmissible, and then issue a warning and suggest users to set the `bounds` option to values other than `"none"`, the default, if the flag is TRUE.

**Comment by yrosseel** [comment]
Hm. My point of view is that if a 'regular' analysis (without any bounds) may result in an inadmissible solution, then it is perfectly all right to allow inadmissible solutions for bootstrap samples. I also see no need to inform the user. Small negative variances may indeed happen, but they just reflect natural variability.

If, on the other hand, a user does not accept an inadmissible solution to begin with, then setting bounds = "standard" may avoid many inadmissible solutions (but not all), and lav_bootstrap_internal() will keep on using these bounds.

**Comment by yrosseel** [comment]
lavaan 0.6-12 will now produce a warning informing the user how many bootstrap runs resulted in an inadmissible solution (both sem/lavaan and bootstrapLavaan)
